### PR TITLE
core(tracehouse): allow child to start <1ms before parent

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Addresses one of the 3 remaining cases for `Fatal trace logic error - child cannot end after parent` error.

**Related Issues/PRs**
#7764, depends on #9785

### CODE REVIEW

File: lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js [hunk 1]
```diff
@@ -430,6 +430,53 @@ describe('Main Thread Tasks', () => {
     ]);
   });
 
+  it('should handle nested tasks of the same name', () => {
+    /*
+    An artistic rendering of the below trace:
+      ████████████████TaskA██████████████████
```
Comment (paulirish) [inline lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js]: s/TaskA/SameName/g

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 2]
```diff
@@ -262,10 +262,36 @@ class MainThreadTasks {
...
             nextTask.duration = nextTask.endTime - nextTask.startTime;
+          } else if (
+            nextTask.startTime - currentTask.startTime < 1000 &&
+            !currentTask.children.length
+          ) {
+            // The parent started less than 1ms before the child, we'll let it slide by swapping the two,
+            // and increasing the duration of the parent.
+            const parentTask = nextTask;
+            const childTask = currentTask;
+            const grandparentTask = currentTask.parent;
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: so `const grandparentTask = childTask.parent;` is *less* confusing?

[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: First i think we need a comment along the lines of:

> Normally, within this sortedTasks loop, our currentTask is oftentimes a parent of nextTask.
> However in this case, the two are swapped. 

then i think different var names like: 
* `parentTask` => `actualParentTask`
* `childTask` => `actualChildTask`

and for grandparent we probably just need a comment:

> Grab the grandparent task and we'll reassign it to be actualParentTask's parent at the end

and yah i'm fine keeping the definition as `currentTask.parent`

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 3]
```diff
@@ -262,10 +262,36 @@ class MainThreadTasks {
             // It's ending at traceEndTs, it means we were missing the end event. We'll truncate it to the parent.
             nextTask.endTime = currentTask.endTime;
             nextTask.duration = nextTask.endTime - nextTask.startTime;
+          } else if (
+            nextTask.startTime - currentTask.startTime < 1000 &&
+            !currentTask.children.length
+          ) {
+            // The parent started less than 1ms before the child, we'll let it slide by swapping the two,
```
Comment (paulirish) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: can you add a graphic like this in here?

```
    ████████████currentTask.parent██████████████████
            █████████nextTask██████████████
           ███████currentTask███████
```

its so great

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: > then i think different var names like:
parentTask => actualParentTask
childTask => actualChildTask

Heh, that's what they were originally :) sgtm

> and for grandparent we probably just need a comment:

sg 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: oooooo, I love it ❤️

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 4]
```diff
@@ -262,10 +262,42 @@ class MainThreadTasks {
...
```
Comment (brendankenny) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: should this be the opposite (parent<->child)?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 5]
```diff
@@ -262,10 +262,42 @@ class MainThreadTasks {
...
+            // and increasing the duration of the parent. Below is an artistic rendition of this situation.
```
Comment (brendankenny) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js]: this is the rendition *after* the swap, right?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: maybe we should give up using 'parent' and 'child' in these swapping contexts 😆 this is written in terms of the actual structure of the trace events, not what we the current object structure might reflect. 

maybe we take the same variable name approach from before and just slap `actual` before these two?

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js]: errr, kind of. once we fix everything `nextTask` and `currentTask` labels will be switched so the diagram looks correct. at this exact moment in the code there is no heirarchy relationship whatsoever between nextTask and currentTask this diagram is attempting to take the true structure of things and apply the variable labels we currently have for them.

open to whatever explanation you think fits best and doesn't confuse me too :)

**Comment by patrickhulce** [comment]
@brendankenny made a few comment tweaks I hope it makes sense!

**Comment by brendankenny** [comment]
> @brendankenny made a few comment tweaks I hope it makes sense!

looks great!

**Pull request merged**